src/requirements/graph/processor/add-opt-out-choices.ts

```diff
@@ -0,0 +1,48 @@
+import GraphProcessor from './interface';
+import RequirementFulfillmentGraph from '..';
+import { CourseForRequirementGraph } from '../types';
+
+export type AddOptOutChoicesParameters<Requirement extends string> = {
+  /**
+   * The mapping from course's unique ID to opt out choices.
+   */
+  readonly userChoiceOnRequirementOverrides: {
+    readonly [uniqueId: string]: {
+      readonly optOut: readonly Requirement[];
+    };
+  };
+};
+
+/**
+ * Add opt out choices.
+ */
+export default class AddOptOutChoices<
+  Requirement extends string,
+  Course extends CourseForRequirementGraph
+> implements GraphProcessor<Requirement, Course> {
+  private addOptOutChoicesParameters: AddOptOutChoicesParameters<Requirement>;
+
+  constructor(addOptOutChoicesParameters: AddOptOutChoicesParameters<Requirement>) {
+    this.addOptOutChoicesParameters = addOptOutChoicesParameters;
+  }
+
+  public process(graph: RequirementFulfillmentGraph<Requirement, Course>) {
+    return addOptOutChoices(graph.copy(), this.addOptOutChoicesParameters);
+  }
+}
+
+export function addOptOutChoices<
+  Requirement extends string,
+  Course extends CourseForRequirementGraph
+>(
+  graph: RequirementFulfillmentGraph<Requirement, Course>,
+  { userChoiceOnRequirementOverrides }: AddOptOutChoicesParameters<Requirement>
+): RequirementFulfillmentGraph<Requirement, Course> {
+  graph.getAllCourses().forEach(course => {
+    userChoiceOnRequirementOverrides[course.uniqueId]?.optOut?.forEach(requirement => {
+      graph.removeEdge(requirement, course);
+    });
+  });
```

Comment:
Can we abstract this pattern somehow? It's repeated almost exactly in `addArbitraryOptInChoices`. Maybe we could define an operation that works on a supertype of `AddOptOutChoicesParameters` and `AddArbitraryOptInChoicesParameters`.

Reply:
I tried adding an abstract class `AddOverriddenChoices` that each of these visitors extended, but it ended up adding more code. It should be okay since we're not planning on adding any more kinds of overridden choices.